lightbluetent/admin.py

```diff
@@ -1,39 +1,194 @@
+def add(short_name):
+    current_crsid = auth_decorator.principal
+    current_user = User.query.filter_by(crsid=current_crsid).first()
+    current_user_socs = current_user.societies
+
+    # Check the society exists
+    society = Society.query.filter_by(uid=short_name).first()
+    society_admins = society.admins
+    if not society:
+        abort(404)
+    # Check that the currently authenticated user is allowed to administer this society.
+    if not society in current_user_socs:
+        abort(403)
+
     if request.method == "POST":
 
-        name =                 request.form["name"]
-        new_short_name =       request.form["short_name"]
-        website =              request.form["website"]
-        description =          request.form["description"]
-        welcome_text =         request.form["welcome_text"]
-        logo =                 request.form["logo"]
-        banner_text =          request.form["banner_text"]
-        banner_color =         request.form["banner_color"]
-        mute_on_start =        request.form["mute_on_start"]
-        disable_private_chat = request.form["disable_private_chat"]
-
-        errors = []
-
-        if not name:
-            errors.append("A society name is required.")
-        if not new_short_name:
-            errors.append("A short society name is required.")
-
-        # TODO: confirm new_short_name is unique
-
-        if not errors:
-            pass
-            # TODO: write to the database
+        values = {}
+
+        for key in ("first_name", "surname", "email_address", "crsid"):
+            values[key] = request.form.get(key, "").strip()
+
+        errors = {}
+
+        # If the user being added is already registered, then simply get that
+        # add this society to that user's list of societies
+
+        # TODO: identifying users should be CRSid-based, changed to email address for testing purposes
+        existing_user = User.query.filter_by(email=values["email_address"]).first()
+        if existing_user:
+            # Check the existing user isn't already an admin
+            if not society in existing_user.societies:
+                pass
+            else:
+                errors["crsid"] = "This user is already an administrator."
+
+            # Check that the values we already have have been given by the registrant.
+            if existing_user.first_name != values["first_name"]:
+                errors["first_name"] = "This CRSid is already registered, but not under this name. Please use the user's registered name."
+            if existing_user.surname != values["surname"]:
+                errors["surname"] = "This CRSid is already registered, but not under this name. Please use the user's registered name."
+            if existing_user.email != values["email_address"]:
+                errors["email_address"] = "This CRSid is already registered, but not under this email address. Please use the user's registered email address."
```

Comment:
Is this a good idea?  I don't really see the point in making sure that someone's first name, surname, email address as they were previously entered into LBT (including typos and extra whitespace) match exactly on addition.

I asked in chat but will ask here too for coverage - what's the reason for checking these if the user exists?  It might be less surprising if we just ignore the details given here if the user already exists.